src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,222 @@
+#include "NuulsUploader.hpp"
+
+#include "common/Env.hpp"
+#include "common/NetworkRequest.hpp"
+#include "providers/twitch/TwitchMessageBuilder.hpp"
+
+#include <QHttpMultiPart>
+
+#define UPLOAD_DELAY 2000
+// Delay between uploads in milliseconds
+
+namespace {
+
+QString getImageFileFormat(const QString &path)
+{
+    static QStringList listOfImageFormats = {".png", ".jpg", ".jpeg"};
+    for (const QString &format : listOfImageFormats)
+    {
+        if (path.endsWith(format, Qt::CaseInsensitive))
+        {
+            return format.mid(1);
+        }
+    }
+    return QString();
+}
+
+boost::optional<QByteArray> convertToPng(QImage image)
+{
+    QByteArray imageData;
+    QBuffer buf(&imageData);
+    buf.open(QIODevice::WriteOnly);
+    bool success = image.save(&buf, "png");
+    if (success)
+    {
+        return boost::optional<QByteArray>(imageData);
+    }
+    else
+    {
+        return boost::optional<QByteArray>(boost::none);
+    }
+}
+}  // namespace
+
+namespace chatterino {
+// These variables are only used from the main thread.
+bool isUploading = false;
+std::queue<RawImageData> uploadQueue;
+
+void uploadImageToNuuls(RawImageData imageData, ChannelPtr channel,
+                        ResizingTextEdit &textEdit)
+{
+    const static char *boundary = "thisistheboudaryasd";
+    const static QString contentType =
+        QString("multipart/form-data; boundary=%1").arg(boundary);
+    static QUrl url(Env::get().imageUploaderUrl);
+
+    QHttpMultiPart *payload = new QHttpMultiPart(QHttpMultiPart::FormDataType);
+    QHttpPart part = QHttpPart();
+    part.setBody(imageData.data);
+    part.setHeader(QNetworkRequest::ContentTypeHeader,
+                   QString("image/%1").arg(imageData.format));
+    part.setHeader(QNetworkRequest::ContentLengthHeader,
+                   QVariant(imageData.data.length()));
+    part.setHeader(
+        QNetworkRequest::ContentDispositionHeader,
+        QString("form-data; name=\"attachment\"; filename=\"control_v.%1\"")
+            .arg(imageData.format));
+    payload->setBoundary(boundary);
+    payload->append(part);
+    NetworkRequest(url, NetworkRequestType::Post)
```

Comment:
I am unsure of current implementation of NetworkRequests, are they executed in the main thread?

Reply:
The current implementation of NetworkRequest should be fully run on separate worker thread(s)

Reply:
Then the part about uploadMutex and uploadQueue being accessed only in main thread does not seem to be true